dali/pipeline/data/tensor_list.cc

```diff
@@ -791,11 +791,16 @@ void TensorList<Backend>::SetContiguity(BatchContiguity state) {
 
 
 template <typename Backend>
-void TensorList<Backend>::MakeContiguous(std::weak_ptr<void> owner) {
+void TensorList<Backend>::MakeContiguous(AccessOrder copy_order) {
   if (state_.IsContiguous()) {
     return;
   }
-  DALI_FAIL("Coalescing the buffer to Contiguous state is not yet implemented.");
+  auto contiguous_tmp = TensorList<Backend>(shape().num_samples());
+  contiguous_tmp.SetContiguity(BatchContiguity::Contiguous);
+  contiguous_tmp.SetupLike(*this);
+  contiguous_tmp.Copy(*this, order());
+  copy_order.wait(order());
+  *this = std::move(contiguous_tmp);
```

Comment:
I am not sure how do I feel about the swap of the allocation, at least to happen automatically for some of the calls. 

If we ever want to allow modification of TensorList memory from Python side, this will prevent things like:
```
tl = TensorListGPU(some_list_of_samples)
t0 = tl[0]  # this will take a reference to memory
t = tl.as_tensor()  # this will make the internal representation contiguous, detaching t0 from tl.
modify(tl)  # the changes are not propagated to t0, as it's separate allocation.
```
With current design `tl` and `t0` would be still aliasing. 

Also, I just realized, that the backend_impl's  "at" is probably broken (and it would even be more broken), as it just takes a raw pointer and wraps it into Python buffer without any reference count subscription. We already had such issues with `__get__`.

Should we get this to open floor to decide what is fine and what is not? Should we just allow non-contiguous TensorList and make the conversion explicit? Should we pursue the current approach?

Also, how much do we want to synchronize the data with host or user provided stream in such cases? 
This mostly impacts the copy_to_external.

Reply:
Sure, let us discuss this further.

> This mostly impacts the copy_to_external.

I think it is not affected. We have a copy there that works no matter if the underlying is continuous or not.